subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/fingerprint/ConfigurationCacheFingerprintChecker.kt

```diff
@@ -62,12 +62,12 @@ class ConfigurationCacheFingerprintChecker(private val host: Host) {
                     }
                 }
                 is ConfigurationCacheFingerprint.ValueSource -> input.run {
-                    checkFingerprintValueIsUpToDate(obtainedValue)?.let { reason ->
+                    checkFingerprintValueIsUpToDate(obtainedValue)?.let<InvalidationReason, InvalidationReason> { reason ->
```

Comment:
💄 I guess the inner `return` prevents Kotlin type inference on the `let` here. In these cases I think it's better to use a regular `val` and an `if `.

```kotlin
val reason = checkFingerprintValueIsUpToDate(...);
if (reason != null) {
    return reason
}
```